# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:   <br>
- Bhuvana Rangaiah Cheemala Marri
- Juan Pablo Salas Suarez

Url: https://github.com/bhuvanarangaiah/03MAIR-Algoritmos-de-Optimizacion-2026<br>
Problema:
> 1. Sesiones de doblaje <br>

Descripción del problema:

- Se precisa coordinar el doblaje de una película. Los actores del doblaje deben coincidir en las
tomas en las que sus personajes aparecen juntos en las diferentes tomas. Los actores de
doblaje cobran todos la misma cantidad por cada día que deben desplazarse hasta el estudio de
grabación independientemente del número de tomas que se graben. No es posible grabar más
de 6 tomas por día. El objetivo es planificar las sesiones por día de manera que el gasto por los
servicios de los actores de doblaje sea el menor posible.<br>
    - Los datos son:<br>
        - Número de actores: 10<br>
        - Número de tomas : 30<br>

Actores/Tomas : https://bit.ly/36D8IuK <br>
- 1 indica que el actor participa en la toma
- 0 en caso contrario

....

(*) La respuesta es obligatoria





                                        

In [1]:
# ============================================================
# CARGA DE DATOS - Problema 1: Sesiones de doblaje
# ============================================================
# Datos: 30 tomas x 10 actores. M[t][a] = 1 si el actor a
# participa en la toma t, 0 en caso contrario.
# Los incrustamos directamente para que el notebook sea
# autocontenido (no depende de ficheros externos en Colab).

MAX_TOMAS_POR_DIA = 6

# Matriz de participacion (formato de entrada, tal cual la hoja)
matriz = [
    [1,1,1,1,1,0,0,0,0,0],  # toma 1
    [0,0,1,1,1,0,0,0,0,0],  # toma 2
    [0,1,0,0,1,0,1,0,0,0],  # toma 3
    [1,1,0,0,0,0,1,1,0,0],  # toma 4
    [0,1,0,1,0,0,0,1,0,0],  # toma 5
    [1,1,0,1,1,0,0,0,0,0],  # toma 6
    [1,1,0,1,1,0,0,0,0,0],  # toma 7
    [1,1,0,0,0,1,0,0,0,0],  # toma 8
    [1,1,0,1,0,0,0,0,0,0],  # toma 9
    [1,1,0,0,0,1,0,0,1,0],  # toma 10
    [1,1,1,0,1,0,0,1,0,0],  # toma 11
    [1,1,1,1,0,1,0,0,0,0],  # toma 12
    [1,0,0,1,1,0,0,0,0,0],  # toma 13
    [1,0,1,0,0,1,0,0,0,0],  # toma 14
    [1,1,0,0,0,0,1,0,0,0],  # toma 15
    [0,0,0,1,0,0,0,0,0,1],  # toma 16
    [1,0,1,0,0,0,0,0,0,0],  # toma 17
    [0,0,1,0,0,1,0,0,0,0],  # toma 18
    [1,0,1,0,0,0,0,0,0,0],  # toma 19
    [1,0,1,1,1,0,0,0,0,0],  # toma 20
    [0,0,0,0,0,1,0,1,0,0],  # toma 21
    [1,1,1,1,0,0,0,0,0,0],  # toma 22
    [1,0,1,0,0,0,0,0,0,0],  # toma 23
    [0,0,1,0,0,1,0,0,0,0],  # toma 24
    [1,1,0,1,0,0,0,0,0,1],  # toma 25
    [1,0,1,0,1,0,0,0,1,0],  # toma 26
    [0,0,0,1,1,0,0,0,0,0],  # toma 27
    [1,0,0,1,0,0,0,0,0,0],  # toma 28
    [1,0,0,0,1,1,0,0,0,0],  # toma 29
    [1,0,0,1,0,0,0,0,0,0],  # toma 30
]

N_TOMAS   = len(matriz)          # 30
N_ACTORES = len(matriz[0])       # 10

# Representacion de trabajo: cada toma -> conjunto de actores.
# (actores numerados 1..10)
tomas = [ {a+1 for a in range(N_ACTORES) if fila[a]==1} for fila in matriz ]

print(tomas)

# --- Verificaciones rapidas ---
print("Nº tomas   :", N_TOMAS)
print("Nº actores :", N_ACTORES)
print("Total participaciones (unos):", sum(sum(f) for f in matriz))
print()
print("Ejemplos (toma -> actores):")
for t in [0,1,15,25]:
    print(f"  toma {t+1:2d}: {sorted(tomas[t])}")
print()
# Cuantas tomas hace cada actor
por_actor = {a: sum(1 for f in matriz if f[a-1]==1) for a in range(1,N_ACTORES+1)}
print("Tomas por actor:", por_actor)


[{1, 2, 3, 4, 5}, {3, 4, 5}, {2, 5, 7}, {8, 1, 2, 7}, {8, 2, 4}, {1, 2, 4, 5}, {1, 2, 4, 5}, {1, 2, 6}, {1, 2, 4}, {1, 2, 6, 9}, {1, 2, 3, 5, 8}, {1, 2, 3, 4, 6}, {1, 4, 5}, {1, 3, 6}, {1, 2, 7}, {10, 4}, {1, 3}, {3, 6}, {1, 3}, {1, 3, 4, 5}, {8, 6}, {1, 2, 3, 4}, {1, 3}, {3, 6}, {1, 2, 10, 4}, {1, 3, 5, 9}, {4, 5}, {1, 4}, {1, 5, 6}, {1, 4}]
Nº tomas   : 30
Nº actores : 10
Total participaciones (unos): 94

Ejemplos (toma -> actores):
  toma  1: [1, 2, 3, 4, 5]
  toma  2: [3, 4, 5]
  toma 16: [4, 10]
  toma 26: [1, 3, 5, 9]

Tomas por actor: {1: 22, 2: 14, 3: 13, 4: 15, 5: 11, 6: 8, 7: 3, 8: 4, 9: 2, 10: 2}


### (*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



### ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Respuesta

In [2]:
# ============================================================
# NUMERO DE POSIBILIDADES - Problema 1: Sesiones de doblaje
# ============================================================
import math

# ---- SIN restricciones -------------------------------------
# Una planificacion = una particion de las 30 tomas en dias
# (cualquier nº de dias, cualquier tamaño). Su numero es el
# NUMERO DE BELL B(30), que calculamos con el "triangulo de Bell"
# (solo sumas: cada fila empieza con el ultimo valor de la
#  anterior; cada entrada = izquierda + diagonal superior-izq;
#  el 1er valor de cada fila es el numero de Bell).
def bell(n):
    fila = [1]                     # B(0) = 1
    for _ in range(n):
        siguiente = [fila[-1]]     # empieza con el ultimo de la fila previa
        for x in fila:
            siguiente.append(siguiente[-1] + x)
        fila = siguiente
    return fila[0]                 # primer elemento = B(n)

B30 = bell(30)

# ---- CON restricciones -------------------------------------
# 30 tomas, maximo 6/dia  ->  exactamente 5 dias de 6 tomas.
# Nº de formas de repartir 30 tomas distintas en 5 grupos
# (no etiquetados) de 6:  30! / ( (6!)^5 * 5! )
con_restr = math.factorial(30) // (math.factorial(6)**5 * math.factorial(5))

# ---- CON la restriccion LITERAL (<=6 tomas/dia, nº de dias libre) ----
# Particiones de 30 tomas en bloques de tamaño <=6 (cualquier nº de dias):
# recurrencia a(n) = sum_{j=1..6} C(n-1, j-1) * a(n-j)
from functools import lru_cache
@lru_cache(None)
def particiones_max6(n):
    if n == 0: return 1
    return sum(math.comb(n-1, j-1) * particiones_max6(n-j)
               for j in range(1, min(6, n) + 1))
total_validas = particiones_max6(30)

print("SIN restricciones  -> B(30) =", B30)
print("                            = {:.3e}".format(B30))
print()
print("CON restriccion (5 dias x 6 tomas):")
print("   30! / (6!^5 * 5!) =", con_restr)
print("                     = {:.3e}".format(con_restr))
print()
print("Restriccion literal (<=6/dia, nº dias libre):")
print("   particiones en bloques <=6 =", total_validas)
print("                              = {:.3e}".format(total_validas))
print()
print("Nota: como cada dia extra obliga a mas desplazamientos (lo confirma")
print("el ILP: con 6..13 dias no existe solucion mejor que 27), el espacio")
print("de busqueda UTIL son las particiones en 5 dias de 6 tomas.")

# Comprobacion del metodo con valores pequeños conocidos:
# B(0..5) = 1, 1, 2, 5, 15, 52
print()
print("Comprobacion B(0..5):", [bell(n) for n in range(6)])


SIN restricciones  -> B(30) = 846749014511809332450147
                            = 8.467e+23

CON restriccion (5 dias x 6 tomas):
   30! / (6!^5 * 5!) = 11423951396577720
                     = 1.142e+16

Restriccion literal (<=6/dia, nº dias libre):
   particiones en bloques <=6 = 726391948970868949621309
                              = 7.264e+23

Nota: como cada dia extra obliga a mas desplazamientos (lo confirma
el ILP: con 6..13 dias no existe solucion mejor que 27), el espacio
de busqueda UTIL son las particiones en 5 dias de 6 tomas.

Comprobacion B(0..5): [1, 1, 2, 5, 15, 52]


### Modelo para el espacio de soluciones<br>
### (*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


**Respuesta**

Conviene distinguir DOS tipos de datos con estructuras distintas:

**A) Datos de entrada (fijos): quien participa en cada toma.**
La entrada llega como una **matriz** 30×10 de 0/1 (formato de la hoja de
calculo). Es comoda para leer y verificar, pero la operacion que mas repetimos
es *"dado un conjunto de tomas de un dia, ¿que actores distintos se necesitan?"*,
es decir la **UNION** de los actores de esas tomas y su **tamaño**. Por eso
convertimos cada toma en un **conjunto (set) de actores** (o equivalentemente un
**bitmask** entero). Asi:
- coste de un dia = tamaño de la union de los sets de sus tomas → directo.
- con bitmask, la union es un simple OR y el conteo un popcount → muy rapido.

Empezamos con la matriz (como se recibe) y **vimos la necesidad de cambiar** a
lista de sets / bitmasks porque la funcion objetivo es esencialmente una union
de conjuntos: la matriz obligaria a recorrer columnas repetidamente, mientras que
el set/bitmask expresa la operacion de forma natural y eficiente. La matriz solo
se usa para cargar; despues se trabaja con los sets.

**B) Datos de la solucion (cambian durante la busqueda): que tomas van cada dia.**
Una planificacion se representa como una **lista de dias**, y cada dia como una
**lista de indices de toma**  (p.ej. `[[0,1,2],[3,4,5],...]`). Es la estructura
que el algoritmo construye y modifica (mover/intercambiar tomas en la busqueda
local), y permite comprobar la restriccion de ≤6 tomas/dia con `len(dia)`.

**Conclusion:** matriz solo como formato de carga → **lista de sets/bitmasks**
para los datos de entrada (operacion clave = union de conjuntos), y **lista de
listas** (dias → tomas) para la solucion que se explora. Es la combinacion que
mejor se adapta a la funcion objetivo y a las restricciones.


In [3]:
# ============================================================
# ESTRUCTURA DE DATOS - demostracion
# ============================================================
# A) Entrada: matriz (carga) -> lista de SETS de actores por toma
#    (equivalente eficiente: BITMASK entero)
ejemplo_matriz = matriz[0]                      # fila de la toma 1
ejemplo_set    = tomas[0]                        # {1,2,3,4,5}
ejemplo_bitmask = 0
for a in tomas[0]:
    ejemplo_bitmask |= (1 << (a-1))
print("Toma 1 como fila de matriz :", ejemplo_matriz)
print("Toma 1 como set de actores :", sorted(ejemplo_set))
print("Toma 1 como bitmask (bin)  :", format(ejemplo_bitmask, '010b'))

# Operacion clave (coste de un dia) = tamaño de la UNION de sets
dia_ejemplo = [0, 1, 5]                           # tomas 1, 2, 6
union_actores = set().union(*[tomas[t] for t in dia_ejemplo])
print("\nDia con tomas 1,2,6 -> actores union:", sorted(union_actores),
      "-> coste dia =", len(union_actores))

# B) Solucion: lista de dias, cada dia una lista de indices de toma
solucion_ejemplo = [[0,1,2,3,4,5],[6,7,8,9,10,11]]
print("\nSolucion (lista de dias -> tomas):", solucion_ejemplo)
print("Tomas en el dia 1:", len(solucion_ejemplo[0]), "(<= 6 OK)")


Toma 1 como fila de matriz : [1, 1, 1, 1, 1, 0, 0, 0, 0, 0]
Toma 1 como set de actores : [1, 2, 3, 4, 5]
Toma 1 como bitmask (bin)  : 0000011111

Dia con tomas 1,2,6 -> actores union: [1, 2, 3, 4, 5] -> coste dia = 5

Solucion (lista de dias -> tomas): [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11]]
Tomas en el dia 1: 6 (<= 6 OK)


### Según el modelo para el espacio de soluciones<br>
### (*)¿Cual es la función objetivo?

### (*)¿Es un problema de maximización o minimización?

Respuesta

In [4]:
# ============================================================
# FUNCION OBJETIVO - Problema 1: Sesiones de doblaje
# ============================================================
# Una SOLUCION (planificacion) se representa como una lista de
# dias; cada dia es una lista de indices de toma (0..29).
#   Ej: [[0,1,2],[3,4]] -> dia1 graba tomas 1,2,3 ; dia2 graba 4,5
#
# COSTE de un dia = nº de actores distintos que deben acudir ese
#   dia = tamaño de la UNION de los conjuntos de actores de sus
#   tomas (cada actor cobra 1 por dia, sin importar cuantas tomas).
# FUNCION OBJETIVO = suma de los costes de todos los dias.
# OBJETIVO: MINIMIZAR ese coste total (total de "actor-dias").

def coste_dia(dia):
    """Actores distintos necesarios en un dia (union de sus tomas)."""
    actores = set()
    for t in dia:
        actores |= tomas[t]
    return len(actores)

def coste(planificacion):
    """Funcion objetivo: total de actor-dias de una planificacion."""
    return sum(coste_dia(dia) for dia in planificacion)

def es_valida(planificacion):
    """Restricciones del problema:
       - ningun dia con mas de 6 tomas
       - las 30 tomas aparecen exactamente una vez."""
    todas = [t for dia in planificacion for t in dia]
    if any(len(dia) > MAX_TOMAS_POR_DIA for dia in planificacion):
        return False
    return sorted(todas) == list(range(N_TOMAS))

# --- Comprobacion sobre una planificacion "naive" (6 en orden) ---
naive = [list(range(i, i+6)) for i in range(0, N_TOMAS, 6)]
print(naive)
print("Planificacion naive valida?", es_valida(naive))
print("Coste por dia:", [coste_dia(d) for d in naive])
print("COSTE TOTAL  :", coste(naive), "actor-dias")


[[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 23], [24, 25, 26, 27, 28, 29]]
Planificacion naive valida? True
Coste por dia: [7, 8, 8, 7, 8]
COSTE TOTAL  : 38 actor-dias


### Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

In [5]:
# ============================================================
# FUERZA BRUTA - Problema 1: Sesiones de doblaje
# ============================================================
# Idea: enumerar TODAS las planificaciones validas (particiones
# de las tomas en dias de <= 6 tomas), evaluar cada una con la
# funcion objetivo 'coste' y quedarnos con la de menor coste.
# Es correcto (garantiza el optimo) pero INTRATABLE para n=30.

def particiones(items, max_size=MAX_TOMAS_POR_DIA):
    """Genera todas las particiones de 'items' en bloques (dias)
       de tamaño <= max_size. Backtracking clasico: cada toma se
       añade a un dia existente (si cabe) o abre un dia nuevo."""
    if not items:
        yield []
        return
    first, rest = items[0], items[1:]
    for part in particiones(rest, max_size):
        for i in range(len(part)):                 # añadir a un dia existente
            if len(part[i]) < max_size:
                yield part[:i] + [part[i] + [first]] + part[i+1:]
        yield [[first]] + part                      # o abrir un dia nuevo

def fuerza_bruta(indices, max_size=MAX_TOMAS_POR_DIA):
    """Devuelve (mejor_planificacion, mejor_coste, nº_evaluadas)."""
    mejor, mejor_coste, evaluadas = None, float('inf'), 0
    for p in particiones(indices, max_size):
        evaluadas += 1
        c = coste(p)
        if c < mejor_coste:
            mejor_coste, mejor = c, p
    return mejor, mejor_coste, evaluadas

# --- Demostracion de correccion sobre un subconjunto pequeño ---
# (con las 30 tomas es inviable: ~10^16 candidatos)
sub = list(range(8))          # primeras 8 tomas
plan, c, n = fuerza_bruta(sub)
print(f"Subconjunto de 8 tomas: {n} particiones evaluadas")
print(f"Mejor coste = {c} actor-dias")
print("Mejor planificacion (tomas 1..8):",
      [[t + 1 for t in dia] for dia in plan])


Subconjunto de 8 tomas: 4131 particiones evaluadas
Mejor coste = 12 actor-dias
Mejor planificacion (tomas 1..8): [[2, 1], [8, 7, 6, 5, 4, 3]]


### Calcula la complejidad del algoritmo por fuerza bruta

**Respuesta**

La complejidad de la fuerza bruta es el producto de dos factores:

**(1) Nº de candidatos a evaluar.** Cada planificacion valida es una particion
de las 30 tomas en dias de <= 6 tomas. Como 30 = 5 × 6 y usar mas de 5 dias
nunca mejora el coste (cada dia extra añade desplazamientos; lo confirma el
ILP), basta explorar las particiones en **5 dias de 6 tomas**, cuyo numero es:

$$\frac{30!}{(6!)^5 \cdot 5!} = 11\,423\,951\,396\,577\,720 \approx 1.14 \times 10^{16}$$

Este numero crece de forma **super-exponencial** con n (del orden del numero de
Bell / Stirling): añadir tomas multiplica el espacio de busqueda por un factor
creciente, por lo que es el termino que domina la complejidad.

**(2) Coste de evaluar un candidato.** La funcion objetivo recorre las n tomas
una vez y, por cada una, une su conjunto de actores (hasta A = 10). Es decir,
**O(n · A)** = O(30 · 10) ≈ 300 operaciones por candidato. Este factor es lineal
y despreciable frente al (1).

**Complejidad total:**

$$O\!\left(\frac{n!}{(6!)^{n/6}\,(n/6)!} \cdot n \cdot A\right) \;\;\approx\;\; O(\text{super-exponencial})$$

El termino combinatorio (nº de particiones) domina; la evaluacion O(n·A) no
cambia la clase de complejidad.

**En tiempo real (con restriccion).** Suponiendo 10⁹ operaciones/segundo:
1.14 × 10¹⁶ candidatos × 300 op ≈ 3.4 × 10¹⁸ operaciones ≈ **~108 años**.

**Conclusion:** aun con la restriccion de 6 tomas/dia, la fuerza bruta es
**intratable** para n = 30. Usar bitmasks o codigo mas rapido solo reduce la
constante (las ~300 op), pero NO el termino super-exponencial: hace falta
cambiar de estrategia (heuristica / metaheuristica), no optimizar el codigo.


In [6]:
# ============================================================
# COMPLEJIDAD FUERZA BRUTA - estimacion de tiempo (con restriccion)
# ============================================================
import math

n, A, max_dia = N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA   # 30, 10, 6
n_dias = n // max_dia                                    # 5 dias de 6 tomas

# Nº de candidatos con la restriccion (5 dias x 6 tomas):
candidatos = math.factorial(n) // (math.factorial(max_dia)**n_dias
                                   * math.factorial(n_dias))
ops_por_candidato = n * A                                # coste O(n·A)
operaciones = candidatos * ops_por_candidato

OPS_POR_SEG = 1e9                                        # 1000 millones op/s
segundos = operaciones / OPS_POR_SEG
anios = segundos / (3600 * 24 * 365)

print(f"Candidatos (5 dias x 6 tomas): {candidatos:.3e}")
print(f"Operaciones por candidato O(n·A): {ops_por_candidato}")
print(f"Operaciones totales: {operaciones:.3e}")
print(f"Tiempo estimado a 1e9 op/s: {segundos:.3e} s  ~  {anios:,.0f} años")


Candidatos (5 dias x 6 tomas): 1.142e+16
Operaciones por candidato O(n·A): 300
Operaciones totales: 3.427e+18
Tiempo estimado a 1e9 op/s: 3.427e+09 s  ~  109 años


### (*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

**Respuesta**

Como el problema es de tipo combinatorio (particion de conjuntos, NP-duro),
la fuerza bruta es intratable. Proponemos un metodo **heuristico** en dos
etapas que encuentra soluciones muy buenas en tiempo polinomico:

**Etapa 1 — Greedy constructivo (clustering).** Cada toma empieza en su propio
dia y se fusionan repetidamente los dos grupos que MENOS actores nuevos añaden
(los que mas comparten), respetando el limite de 6 tomas/dia, hasta quedar 5
dias. Aplica directamente la intuicion del problema: juntar tomas que comparten
actores. Coste: **O(n³·A)**.

**Etapa 2 — Busqueda local (refinamiento).** Sobre la solucion greedy se aplican
pequeños cambios (MOVER una toma a otro dia, o INTERCAMBIAR dos tomas entre
dias) y se aceptan si no empeoran el coste (hill-climbing). Con reinicios
aleatorios se evitan optimos locales. Cada iteracion es O(n·A).

**Por que mejora a la fuerza bruta.** La fuerza bruta es super-exponencial
(numero de Bell ~10¹⁶–10²³) y tarda siglos. La heuristica NO enumera todas las
particiones: construye una y la refina, con complejidad **polinomica**, y
termina en milisegundos. El precio es que no garantiza el optimo, pero:

- Pasa de **38** (naive) a **27** actor-dias.
- La cota inferior teorica es 21.
- El ILP (metodo exacto, seccion posterior) demuestra que **27 es el optimo
  global**, es decir, la heuristica alcanza el optimo real del problema.


In [7]:
import itertools, random

# ============================================================
# ALGORITMO MEJORADO (HEURISTICO) - Problema 1: Sesiones de doblaje
# Dos etapas:  (1) Greedy constructivo  ->  (2) Busqueda local
# ============================================================

# ---- ETAPA 1: GREEDY CONSTRUCTIVO (clustering por solapamiento) ----
# Cada toma empieza en su propio dia; fusionamos repetidamente los
# dos grupos cuya union añade menos actores nuevos (mas comparten),
# sin superar 6 tomas/dia, hasta quedarnos con 5 dias.
def greedy_constructivo(ndias=5):
    grupos = [[t] for t in range(N_TOMAS)]
    while len(grupos) > ndias:
        mejor = None                                   # (delta, i, j)
        for i, j in itertools.combinations(range(len(grupos)), 2):
            if len(grupos[i]) + len(grupos[j]) <= MAX_TOMAS_POR_DIA:
                fusion = grupos[i] + grupos[j]
                delta = coste_dia(fusion) - coste_dia(grupos[i]) - coste_dia(grupos[j])
                if mejor is None or delta < mejor[0]:
                    mejor = (delta, i, j)
        if mejor is None:
            break
        _, i, j = mejor
        grupos[i] = grupos[i] + grupos[j]
        del grupos[j]
    return grupos

# ---- ETAPA 2: BUSQUEDA LOCAL (refinamiento, hill-climbing) ----
# Pequeños cambios sobre la solucion: MOVER una toma a otro dia o
# INTERCAMBIAR dos tomas entre dias. Se acepta si no empeora el
# coste. Con varios reinicios escapamos de optimos locales.
def busqueda_local(dias, iters=20000, seed=0):
    random.seed(seed)
    dias = [d[:] for d in dias]
    best = coste(dias)
    for _ in range(iters):
        if random.random() < 0.5:                      # MOVER
            d1, d2 = random.sample(range(len(dias)), 2)
            if not dias[d1] or len(dias[d2]) >= MAX_TOMAS_POR_DIA:
                continue
            i = random.randrange(len(dias[d1]))
            t = dias[d1].pop(i); dias[d2].append(t)
            c = coste(dias)
            if c <= best: best = c
            else: dias[d2].pop(); dias[d1].insert(i, t)         # deshacer
        else:                                          # INTERCAMBIAR
            d1, d2 = random.sample(range(len(dias)), 2)
            if not dias[d1] or not dias[d2]:
                continue
            i = random.randrange(len(dias[d1])); j = random.randrange(len(dias[d2]))
            dias[d1][i], dias[d2][j] = dias[d2][j], dias[d1][i]
            c = coste(dias)
            if c <= best: best = c
            else: dias[d1][i], dias[d2][j] = dias[d2][j], dias[d1][i]  # deshacer
    return dias, best

# ---- HEURISTICA COMPLETA ----
def resolver(reinicios=10):
    inicial = greedy_constructivo()
    mejor_sol, mejor_c = inicial, coste(inicial)
    for s in range(reinicios):
        sol, c = busqueda_local(inicial, seed=s)
        if c < mejor_c:
            mejor_c, mejor_sol = c, sol
    return mejor_sol, mejor_c

sol, c = resolver()
print("Etapa 1 (greedy)      :", coste(greedy_constructivo()), "actor-dias")
print("Etapa 2 (busqueda local):", c, "actor-dias   valida?", es_valida(sol))
print("(naive=38 | cota inferior=21 | optimo global=27 demostrado por ILP)")
for k, d in enumerate(sol, 1):
    ac = sorted(set().union(*[tomas[t] for t in d]))
    print(f"  Dia {k}: tomas {sorted(t+1 for t in d)} -> {len(ac)} actores {ac}")


Etapa 1 (greedy)      : 31 actor-dias
Etapa 2 (busqueda local): 27 actor-dias   valida? True
(naive=38 | cota inferior=21 | optimo global=27 demostrado por ILP)
  Dia 1: tomas [1, 5, 9, 11, 20, 30] -> 6 actores [1, 2, 3, 4, 5, 8]
  Dia 2: tomas [7, 13, 16, 25, 27, 28] -> 5 actores [1, 2, 4, 5, 10]
  Dia 3: tomas [3, 4, 8, 15, 21, 29] -> 6 actores [1, 2, 5, 6, 7, 8]
  Dia 4: tomas [2, 6, 10, 12, 22, 26] -> 7 actores [1, 2, 3, 4, 5, 6, 9]
  Dia 5: tomas [14, 17, 18, 19, 23, 24] -> 3 actores [1, 3, 6]


### (*)Calcula la complejidad del algoritmo

**Respuesta**

Complejidad del algoritmo mejorado (heuristico), en dos etapas:

**Etapa 1 — Greedy constructivo.** Realiza ~(n − 5) fusiones. En cada fusion
examina todos los pares de grupos, O(n²), y evalua cada union en O(A). Coste:

$$O(n^3 \cdot A)$$

**Etapa 2 — Busqueda local.** Con `I` iteraciones; en cada una hace un
movimiento (mover/intercambiar) y recalcula el coste en O(n · A). Coste:

$$O(I \cdot n \cdot A)$$

**Total: polinomico**, dominado por O(n³·A) del greedy (I es una constante
fijada). Frente a la fuerza bruta (super-exponencial, numero de Bell ~10¹⁶–10²³)
o a los metodos exactos (DP O(2ⁿ), ILP exponencial), la heuristica es
**polinomica** y termina en milisegundos, a cambio de no garantizar el optimo
(aunque en la practica lo alcanza: 27 actor-dias, confirmado por el ILP).


In [8]:
# ============================================================
# COMPLEJIDAD DEL ALGORITMO MEJORADO - comprobacion empirica
# ============================================================
# Medimos el tiempo del greedy al crecer n para ver el comportamiento
# polinomico (muy lejos del super-exponencial de la fuerza bruta).
# Generador minimo e INLINE (no depende de celdas posteriores).
import time, math, random
_orig = (matriz, tomas, N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA)

print(" n  | tiempo greedy (s)")
for nt in [30, 60, 120, 240]:
    rng = random.Random(7)
    tks = [set(rng.sample(range(1, 13), rng.randint(1, 5))) for _ in range(nt)]
    tomas = tks
    N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA = nt, 12, 6
    t0 = time.time(); greedy_constructivo(math.ceil(nt/6)); dt = time.time()-t0
    print(f" {nt:3d} | {dt:.3f}")

matriz, tomas, N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA = _orig   # restaurar
print("(crecimiento polinomico O(n^3*A), no exponencial)")


 n  | tiempo greedy (s)
  30 | 0.003
  60 | 0.020
 120 | 0.154


 240 | 1.235
(crecimiento polinomico O(n^3*A), no exponencial)


## Enfoque alternativo (EXACTO): Programación Lineal Entera (ILP)

Ademas de la heuristica, el problema se puede modelar como un **problema de
Programacion Lineal Entera** (variables binarias 0/1), analogo al problema de
inversion de capital visto en clase. Un solver (CBC) devuelve el **optimo
demostrado**.

Fijamos **D = 5 dias** (= 30/6, el minimo posible). Esta eleccion no pierde
el optimo: se ha comprobado resolviendo el ILP con D = 6…13 dias (y con ≥14
dias el coste es ≥ 2·14 = 28 > 27, pues cada dia no vacio requiere ≥2 actores):
ninguna planificacion con mas de 5 dias baja de 27.

**Variables:** `x[t,d]=1` si la toma t se graba el dia d;  `y[a,d]=1` si el
actor a debe acudir el dia d.

**Objetivo:** minimizar los actor-dias  →  min Σ y[a,d]

**Restricciones:**
1. cada toma en exactamente un dia:  Σ_d x[t,d] = 1
2. maximo 6 tomas por dia:  Σ_t x[t,d] ≤ 6
3. enlace: si la toma t (con actor a) va el dia d, el actor acude:  y[a,d] ≥ x[t,d]

**Comparativa de los dos enfoques** (ambos alcanzan **27 actor-dias**):

| Enfoque | Coste | ¿Garantiza optimo? | Complejidad | Tiempo | Escala a n grande |
|---|---|---|---|---|---|
| Heuristico (greedy + busqueda local) | 27 | No (pero lo alcanza) | Polinomica O(n³·A) | milisegundos | Si |
| ILP exacto (CBC) | 27 | **Si (lo demuestra)** | Exponencial (NP-duro) | ~segundos | No |

**Justificacion:** el ILP **demuestra** que 27 es el minimo, luego confirma que
la heuristica encuentra el optimo real. Como ambos dan el mismo coste (27), se
elige la **heuristica** como solucion practica (mismo coste, mucho mas rapida y
escalable), usando el ILP como validacion de optimalidad.


In [9]:
# ============================================================
# ENFOQUE ALTERNATIVO (EXACTO): PROGRAMACION LINEAL ENTERA (ILP)
# Devuelve el OPTIMO DEMOSTRADO usando el solver CBC (via PuLP).
# ============================================================
# En Google Colab PuLP viene instalado; si no:  !pip install pulp
import pulp, time

D = 5   # 5 dias x 6 tomas = 30

prob = pulp.LpProblem("doblaje", pulp.LpMinimize)
# x[t,d]=1 si la toma t se graba el dia d ; y[a,d]=1 si el actor a acude el dia d
x = {(t,d): pulp.LpVariable(f"x_{t}_{d}", cat="Binary")
     for t in range(N_TOMAS) for d in range(D)}
y = {(a,d): pulp.LpVariable(f"y_{a}_{d}", cat="Binary")
     for a in range(N_ACTORES) for d in range(D)}

# Objetivo: minimizar actor-dias
prob += pulp.lpSum(y[a,d] for a in range(N_ACTORES) for d in range(D))

# (1) cada toma en exactamente un dia
for t in range(N_TOMAS):
    prob += pulp.lpSum(x[t,d] for d in range(D)) == 1
# (2) maximo 6 tomas por dia
for d in range(D):
    prob += pulp.lpSum(x[t,d] for t in range(N_TOMAS)) <= MAX_TOMAS_POR_DIA
# (3) enlace actor-dia: si la toma t (con actor a) va el dia d, el actor acude
for t in range(N_TOMAS):
    for a in range(N_ACTORES):
        if matriz[t][a] == 1:
            for d in range(D):
                prob += y[a,d] >= x[t,d]

# Rompe-simetria: ordena los dias por su toma de menor indice
# (evita explorar los 5! ordenes equivalentes de dias -> mucho mas rapido)
for d in range(1, D):
    for t in range(N_TOMAS):
        prob += x[t,d] <= pulp.lpSum(x[tp,d-1] for tp in range(t))

t0 = time.time()
prob.solve(pulp.PULP_CBC_CMD(msg=0, timeLimit=60))
print("Estado del solver:", pulp.LpStatus[prob.status])
print("OPTIMO DEMOSTRADO =", int(round(pulp.value(prob.objective))), "actor-dias")
print("Tiempo:", round(time.time()-t0, 1), "s")
for d in range(D):
    tk = sorted(t+1 for t in range(N_TOMAS) if pulp.value(x[t,d]) > 0.5)
    ac = sorted(a+1 for a in range(N_ACTORES) if pulp.value(y[a,d]) > 0.5)
    print(f"  Dia {d+1}: tomas {tk} -> {len(ac)} actores {ac}")


Estado del solver: Optimal
OPTIMO DEMOSTRADO = 27 actor-dias
Tiempo: 23.6 s
  Dia 1: tomas [1, 2, 5, 11, 20, 26] -> 7 actores [1, 2, 3, 4, 5, 8, 9]
  Dia 2: tomas [3, 4, 10, 15, 21, 29] -> 7 actores [1, 2, 5, 6, 7, 8, 9]
  Dia 3: tomas [6, 7, 9, 13, 27, 28] -> 4 actores [1, 2, 4, 5]
  Dia 4: tomas [8, 12, 16, 22, 25, 30] -> 6 actores [1, 2, 3, 4, 6, 10]
  Dia 5: tomas [14, 17, 18, 19, 23, 24] -> 3 actores [1, 3, 6]


### Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

**Respuesta**

Un juego de datos de entrada del problema es una **matriz de participacion
actor×toma** (0/1). El generador `generar_datos` crea instancias aleatorias
**realistas** y parametrizables:

- Nº de tomas y de actores configurables (por defecto 30 y 10, como el problema).
- Cada toma tiene entre `min_act` y `max_act` actores, y **nunca** queda vacia.
- Los actores tienen **popularidad desigual** (unos participan en muchas mas
  tomas que otros), igual que en los datos reales, donde hay protagonistas que
  aparecen casi siempre y secundarios con pocas tomas.
- `seed` para reproducibilidad.

Esto permite **probar el algoritmo con distintos tamaños y densidades** y
comprobar que sigue funcionando (validez de la solucion y coste razonable) mas
alla del juego de datos concreto del enunciado.


In [10]:
# ============================================================
# JUEGO DE DATOS ALEATORIO - Problema 1: Sesiones de doblaje
# ============================================================
import random

def generar_datos(n_tomas=30, n_actores=10, min_act=1, max_act=5,
                  max_por_dia=6, seed=None):
    """Genera un problema aleatorio realista:
       - cada toma tiene entre min_act y max_act actores (>=1, nunca vacia)
       - los actores tienen POPULARIDAD distinta (unos salen mas que otros),
         como en los datos reales (p.ej. un protagonista en casi todas).
       Devuelve (matriz, tomas_sets, max_por_dia)."""
    rng = random.Random(seed)
    # peso de popularidad por actor (decreciente) -> reparto desigual realista
    pesos = [n_actores - a for a in range(n_actores)]   # actor 1 el mas popular
    matriz, tomas = [], []
    for _ in range(n_tomas):
        k = rng.randint(min_act, max_act)               # nº actores en la toma
        elegidos = set(rng.choices(range(n_actores), weights=pesos, k=k))
        if not elegidos:                                # garantizar no vacia
            elegidos = {rng.randrange(n_actores)}
        fila = [1 if a in elegidos else 0 for a in range(n_actores)]
        matriz.append(fila)
        tomas.append({a+1 for a in elegidos})
    return matriz, tomas, max_por_dia

# --- Generar un juego de ejemplo (mismo tamaño que el problema) ---
m_rnd, tomas_rnd, maxd = generar_datos(n_tomas=30, n_actores=10, seed=42)
print("Juego aleatorio generado: 30 tomas, 10 actores, max 6/dia (seed=42)")
print("Actores por toma (primeras 10):", [len(t) for t in tomas_rnd[:10]])
part = sum(len(t) for t in tomas_rnd)
print("Participaciones totales:", part, "| media actores/toma:", round(part/30,2))
por_actor = {a: sum(1 for t in tomas_rnd if a in t) for a in range(1,11)}
print("Tomas por actor:", por_actor)


Juego aleatorio generado: 30 tomas, 10 actores, max 6/dia (seed=42)
Actores por toma (primeras 10): [1, 2, 4, 1, 4, 2, 2, 1, 1, 2]
Participaciones totales: 71 | media actores/toma: 2.37
Tomas por actor: {1: 12, 2: 12, 3: 8, 4: 10, 5: 8, 6: 8, 7: 6, 8: 2, 9: 2, 10: 3}


### Aplica el algoritmo al juego de datos generado

**Respuesta**

Aplicamos la heuristica (greedy + busqueda local) a instancias generadas por
`generar_datos`. Como las funciones usan las globales del problema, la funcion
`cargar_instancia` las sustituye por la instancia aleatoria antes de resolver
(y al final se restauran los datos originales).

Resultados: en todos los casos la solucion es **valida** (≤6 tomas/dia y todas
las tomas asignadas) y el coste queda por encima —pero cerca— de la cota
inferior teorica, tanto en la instancia de ejemplo como al variar tamaño y
densidad (18×8, 30×10, 48×12, 60×15). Esto confirma que el algoritmo **generaliza**
mas alla del juego de datos del enunciado.


In [11]:
# ============================================================
# APLICAR LA HEURISTICA AL JUEGO DE DATOS ALEATORIO
# ============================================================
# Las funciones (coste, greedy_constructivo, resolver...) usan las
# variables globales del problema. Para resolver una instancia
# aleatoria, cargamos sus datos en esas globales y llamamos a resolver().
import math, copy

# Guardar los datos ORIGINALES del problema para restaurarlos al final
_orig = (matriz, tomas, N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA)

def cargar_instancia(m, tks, maxd):
    global matriz, tomas, N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA
    matriz, tomas = m, tks
    N_TOMAS, N_ACTORES = len(m), len(m[0])
    MAX_TOMAS_POR_DIA = maxd

def resolver_instancia(n_tomas, n_actores, seed):
    m, tks, maxd = generar_datos(n_tomas, n_actores, seed=seed)
    cargar_instancia(m, tks, maxd)
    sol, c = resolver(reinicios=10)              # greedy + busqueda local
    lb = sum(math.ceil(sum(1 for f in m if f[a]==1)/maxd) for a in range(n_actores))
    return sol, c, lb

# --- 1) Resolver la instancia de ejemplo (seed=42, 30 tomas) ---
sol, c, lb = resolver_instancia(30, 10, seed=42)
print("Instancia aleatoria seed=42 (30 tomas, 10 actores):")
print(f"  Coste heuristica = {c} actor-dias | cota inferior = {lb} | valida = {es_valida(sol)}")
for k, d in enumerate(sol, 1):
    ac = sorted(set().union(*[tomas[t] for t in d])) if d else []
    print(f"    Dia {k}: tomas {sorted(t+1 for t in d)} -> {len(ac)} actores {ac}")

# --- 2) Probar con varios tamaños y semillas (robustez) ---
print("\nPruebas con distintos tamaños/semillas:")
print(" tomas x actores | seed | coste | cota inf | valida")
for (nt, na, sd) in [(18,8,1),(30,10,2),(48,12,3),(60,15,4)]:
    sol, c, lb = resolver_instancia(nt, na, sd)
    print(f"   {nt:2d} x {na:2d}       |  {sd}   |  {c:3d}  |   {lb:3d}    | {es_valida(sol)}")

# --- Restaurar los datos ORIGINALES del problema ---
matriz, tomas, N_TOMAS, N_ACTORES, MAX_TOMAS_POR_DIA = _orig
print("\n(datos originales del problema restaurados)")


Instancia aleatoria seed=42 (30 tomas, 10 actores):
  Coste heuristica = 25 actor-dias | cota inferior = 16 | valida = True
    Dia 1: tomas [1, 8, 16, 27, 28, 30] -> 4 actores [1, 5, 7, 9]
    Dia 2: tomas [2, 4, 7, 10, 23, 29] -> 4 actores [1, 2, 4, 6]
    Dia 3: tomas [3, 15, 18, 24, 25, 26] -> 6 actores [1, 2, 3, 4, 5, 10]
    Dia 4: tomas [6, 11, 12, 17, 21, 22] -> 6 actores [1, 2, 4, 5, 7, 8]
    Dia 5: tomas [5, 9, 13, 14, 19, 20] -> 5 actores [2, 3, 6, 7, 10]

Pruebas con distintos tamaños/semillas:
 tomas x actores | seed | coste | cota inf | valida


   18 x  8       |  1   |   17  |    13    | True


   30 x 10       |  2   |   25  |    18    | True


   48 x 12       |  3   |   42  |    26    | True


   60 x 15       |  4   |   61  |    35    | True

(datos originales del problema restaurados)


### Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

**Respuesta**

**Bibliografia de la asignatura**

- Brassard, G. y Bratley, P. (1997). *Fundamentos de algoritmia*. Prentice Hall. ISBN 9788489660007.
- Guerequeta, R. y Vallecillo, A. (2000). *Tecnicas de diseño de algoritmos*. Universidad de Malaga.
- Lee, R. C. T., Tseng, S. S., Chang, R. C. y Tsai, Y. T. (2005). *Introduccion al diseño y analisis de algoritmos*. ISBN 9789701061244.
- Duarte, A. (2008). *Metaheuristicas*. Madrid: Dykinson.
- Material y videoconferencias de la asignatura *Algoritmos de Optimizacion* (VIU).

**Tecnicas y conceptos**

- Numeros de Bell / particiones de conjuntos y numeros de Stirling de segunda especie (para el conteo de posibilidades).
- Heuristicas constructivas (greedy / clustering aglomerativo) y busqueda local (hill-climbing) como metaheuristica de mejora.
- Ramificacion y poda (branch & bound) y programacion dinamica sobre subconjuntos como metodos exactos.
- Programacion Lineal Entera (formulacion con variables binarias 0/1), analoga al problema de inversion de capital visto en clase.

**Herramientas y librerias**

- Python 3 (estructuras `set`, operaciones de bits/bitmask, modulo `random`).
- PuLP (modelado de PL entera) con el solver CBC (COIN-OR Branch and Cut).
- Documentacion de PuLP: https://coin-or.github.io/pulp/

*(Nota: revisa/ajusta esta lista para reflejar unicamente las fuentes que
realmente hayas consultado.)*


### Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño


**Respuesta**

**1. Mejoras algoritmicas.**
La busqueda local se puede reforzar con metaheuristicas mas potentes:
**recocido simulado** (acepta empeoramientos controlados para escapar de optimos
locales), **busqueda tabu** (memoria de movimientos recientes), **algoritmos
geneticos** (poblacion de planificaciones que se cruzan y mutan) o **colonia de
hormigas**. Tambien se pueden usar vecindarios mas ricos (mover bloques de
tomas, reasignar dias completos) y multiarranque para mayor robustez.

**2. Metodos exactos mas escalables.**
Para certificar el optimo en instancias mayores: mejores **cotas inferiores**
para el branch & bound, formulaciones ILP mas ajustadas con **rompe-simetrias**
y **planos de corte**, resolucion con solvers comerciales (Gurobi, CPLEX) o
tecnicas de **generacion de columnas**, dado que el problema es una particion de
conjuntos.

**3. Variaciones al alza del tamaño.**
Al crecer n, los metodos exactos (fuerza bruta, DP O(2^n), ILP) dejan de ser
viables, pero la **heuristica polinomica O(n³·A) escala** y sigue dando buenas
soluciones en milisegundos/segundos. Para tamaños muy grandes se puede
**paralelizar** la busqueda (varios arranques a la vez) y usar bitmasks para
acelerar la evaluacion del coste.

**4. Variaciones del problema (mas realistas).**
- **Cachés distintos por actor**: minimizar el coste ponderado (no todos cobran
  igual) → basta cambiar la funcion objetivo.
- **Disponibilidad de actores**: cada actor solo puede ciertos dias → nuevas
  restricciones.
- **Duracion de las tomas / limite por horas** en vez de nº fijo de tomas/dia.
- **Varios estudios o platos** en paralelo.
- **Precedencias** entre tomas (unas antes que otras) o agrupaciones obligatorias.
- **Objetivo multiple**: minimizar coste y a la vez el nº de dias o los
  desplazamientos (optimizacion multiobjetivo).

En resumen: la estructura (particion de conjuntos con funcion objetivo de union)
se mantiene, y la mayoria de variaciones se modelan añadiendo restricciones o
cambiando la funcion de coste, reutilizando el mismo esquema heuristico/exacto.
